### PR TITLE
Remove obsolete enable/disable module on mobile feature, remove mobile_theme feature remains

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Removes functionality of enabling/disabling a module on mobile devices. Removes remains of mobile_theme feature.
| Type?             | refacto
| Category?         | CO
| BC breaks?        | yes - partial, functionality was removed, but public functions kept with a deprecation notice. They return sensible data. If you ask if a module is `enabled on mobile`, it will answer you with the general `enabled`/`disabled`.
| Deprecations?     | yes - `enable_device` column is no longer used, `mobile_theme` column is no longer used
| How to test?      | Play around with modules, see nothing is broken.
| Fixed ticket?     | 
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/659
| Sponsor company   |

### CODE REVIEW

File: classes/Context.php [hunk 1]
```diff
@@ -212,37 +212,7 @@ public function isTablet()
...
-                            }
-
-                            break;
-                    }
-                }
-            }
-        }
-
-        return $this->mobile_device;
+        return false;
```
Comment (eternoendless) [inline classes/Context.php:226]: Why not deprecate this?

File: classes/Context.php [hunk 2]
```diff
@@ -284,28 +257,7 @@ public function getCurrentLocale()
...
-                $guest->update();
-            }
-        }
-
-        return isset($_SERVER['HTTP_USER_AGENT'], Context::getContext()->cookie)
-            && (bool) Configuration::get('PS_ALLOW_MOBILE_DEVICE')
-            && defined('_PS_THEME_MOBILE_DIR_')
-            && @filemtime(_PS_THEME_MOBILE_DIR_)
-            && !Context::getContext()->cookie->no_mobile;
+        return false;
```
Comment (eternoendless) [inline classes/Context.php:277]: Remember to throw a deprecation error on deprecated methods: https://devdocs.prestashop-project.org/8/development/coding-standards/#deprecations

File: src/Adapter/Module/Module.php [hunk 3]
```diff
@@ -230,9 +227,13 @@ public function isActive(): bool
...
     }
 
+    /**
+     * @deprecated since 9.0.0 - This functionality was disabled. Function will be completely removed
+     * in the next major. There is no replacement, all clients should have the same experience.
+     */
     public function isActiveOnMobile(): bool
     {
-        return (bool) $this->database->get('active_on_mobile');
+        return (bool) $this->database->get('active');
```
Comment (eternoendless) [inline src/Adapter/Module/Module.php]: Should redirect to isActive() instead

File: src/Adapter/Module/ModuleDataProvider.php [hunk 4]
```diff
@@ -104,7 +103,7 @@ public function findByName($name)
         if ($result) {
             $result['installed'] = 1;
             $result['active'] = $this->isEnabled($name);
-            $result['active_on_mobile'] = (bool) ($this->getDeviceStatus($name) & AddonListFilterDeviceStatus::DEVICE_MOBILE);
+            $result['active_on_mobile'] = $this->isEnabled($name);
```
Comment (eternoendless) [inline src/Adapter/Module/ModuleDataProvider.php]: IMO this should be removed

File: src/Core/Module/ModuleManager.php [hunk 5]
```diff
@@ -254,46 +254,22 @@ public function disable(string $name): bool
         return $disabled;
     }
 
+    /**
+     * @deprecated since 9.0.0 - This functionality was disabled. Function will be completely removed
+     * in the next major. There is no replacement, all clients should have the same experience.
+     */
     public function enableMobile(string $name): bool
```
Comment (eternoendless) [inline src/Core/Module/ModuleManager.php]: IMO this could be removed entirely

File: src/Core/Module/ModuleManagerInterface.php [hunk 6]
```diff
@@ -47,8 +47,16 @@ public function enable(string $name): bool;
 
     public function disable(string $name): bool;
 
+    /**
+     * @deprecated since 9.0.0 - This functionality was disabled. Function will be completely removed
+     * in the next major. There is no replacement, all clients should have the same experience.
+     */
     public function enableMobile(string $name): bool;
```
Comment (eternoendless) [inline src/Core/Module/ModuleManagerInterface.php]: Again to be removed

**Comment by eternoendless** [comment]
I'm not completely sure if this feature should be kept for the rare cases where you wouldn't want to enable modules in mobile devices or not. I'd like to hear people's thoughts on this.

File: classes/Context.php (hunk 1 above)
Comment (Hlavtox) [reply classes/Context.php:226]: Fixed

File: classes/Context.php (hunk 2 above)
Comment (Hlavtox) [reply classes/Context.php:277]: Fixed

File: src/Adapter/Module/ModuleDataProvider.php (hunk 4 above)
Comment (Hlavtox) [reply src/Adapter/Module/ModuleDataProvider.php]: Removed

File: src/Core/Module/ModuleManagerInterface.php (hunk 6 above)
Comment (Hlavtox) [reply src/Core/Module/ModuleManagerInterface.php]: Added deprecation error now

File: src/Core/Module/ModuleManager.php (hunk 5 above)
Comment (Hlavtox) [reply src/Core/Module/ModuleManager.php]: Added deprecation error now

File: src/Adapter/Module/Module.php (hunk 3 above)
Comment (Hlavtox) [reply src/Adapter/Module/Module.php]: Done

**Comment by kpodemski** [comment]
There are some edge-case scenarios where we don't want to show modules on mobile/desktop. There are a few issues with this mechanism, tho:
- Cache is not respecting devices. One cache for all of them.
- Checking if the device is a mobile device or not is not working correctly sometimes.

I'm ok with removing it.

cc @PrestaEdit @marsaldev @iqit-commerce @prestarocket @jf-viguier @PrestaSafe @Oksydan @SharakPL @lmeyer1 @rmilecki 

Any thoughts? :)

**Comment by iqit-commerce** [comment]
I had few users who use that possibility, but amount of them was low and I would say equal to users who used that function by mistake :)

**Comment by jf-viguier** [comment]
Remove it

**Comment by Oksydan** [comment]
I am gonna be against it. Actually we are using this feature frequently. 
Separating modules views and options by device is something we are doing in every prestashop store.
Yes. It requires doubling cache by including extra parameter to cache key in modules. If this feature will be removed we have to handle it by our self's in modules configuration. 
It is all about performance. Reducing html initial sizes is important for mobile device to reduce FCP and TBT.

**Comment by Hlavtox** [comment]
@Oksydan Can you provide a specific scenario where a desktop user should be provided a different html than the mobile client? Everybody should get the same experience regardless their device. If I am on a big ipad with a keyboard, is it still a mobile device or a desktop? When I turn a device landscape/portrait, my experience should change, right?

**Comment by rmilecki** [comment]
Let me ask about one possible issue with this change regarding `ps_categorytree` and Classic theme.

Some shop owners don't want categories list to be displayed on **mobile** devices. It's because those trees can be long and that forces customer to scroll a lot before he can see actual products. So some shop owners want categories tree hidden and have customers browse categories using main menu.

Years ago it was handled by Classic theme having `hidden-sm-down` CSS class for categories tree block.
Later it was changed with the commit [04a9b7eb31cd ("Fix categorytree not showing on mobile when mobile is enabled")](https://github.com/PrestaShop/classic-theme/commit/04a9b7eb31cdeee50537931f26db8f8d5bd7b5d4) - I believe the reasoning was that "Disable Mobile" can be used.

If we remove "Disable Mobile" option then there will be no way of hiding categories tree on mobile. I believe the only option will be a custom theme providing `ps_categorytree.tpl`.

Is that OK / acceptable?
I'm not sure if shop owners will be happy having to look for a custom theme for such a simple change.

**Comment by kpodemski** [comment]
@rmilecki 

It's a legit use case, although I think it should be handled on the theme/module template file level, not with the "Disable on mobile" feature.

Whether we remove it or improve cache handling, it would be easier to handle multi-device cache because, right now, this feature doesn't work correctly. I'm against creating more cache contexts and producing thousands of additional files.

**Comment by rmilecki** [comment]
> It's a legit use case, although I think it should be handled on the theme/module template file level, not with the "Disable on mobile" feature.

Understood, thank you. Further discussion on that. should be probably handled out of this pull request.

**Comment by micka-fdz** [comment]
I agree to remove this (old school) feature.

**Comment by SharakPL** [comment]
Most of the traffic is from mobile so it should be mobile first approach anyway. But relying only on css media queries is not optimal. IMO there should be an option to enable/disable modules on different sizes. Maybe we could make use of this package https://github.com/tattali/MobileDetectBundle in the core and allow different templates used on different devices.

**Review by micka-fdz — DISMISSED** [review]

**Comment by nesrineabdmouleh** [comment]
UI tests => https://github.com/nesrineabdmouleh/testing_pr/actions/runs/5118953651

**Review by kpodemski — DISMISSED** [review]

**Review by nesrineabdmouleh — DISMISSED** [review]

**Comment by kpodemski** [comment]
Hi @nesrineabdmouleh 

I see two conflicted files from your tests

**Review by kpodemski — APPROVED** [review]

**Review by nesrineabdmouleh — APPROVED** [review]

**Review by FabienPapet — APPROVED** [review]

**Review by hibatallahAouadni — APPROVED** [review]
Hello @Hlavtox 

As you mentioned the only modules that their enable/disable doesn't work are `ps_banner`, `ps_imageslider` and `ps_themecusto`, otherwise LGTM, QA ✅ 

[pr_31151.webm](https://github.com/PrestaShop/PrestaShop/assets/23308427/c794f488-3f41-4dbc-b139-98ce66fb2281)

Tests ok too ✅ 
https://github.com/aniszr/ga.tests.ui.pr/actions/runs/5154502535

Thanks!

**Comment by kpodemski** [comment]
thanks everyone :)

**Comment by ShaiMagal** [comment]
I think, this is not good idea.
I know, it is "responsive times" now, and "same user experiences" bla bla bla, but sometimes you would like to disable slider on mobile for example.

I don't understand, why you remove it...

Customers are often unhappy, if they upgrade for newer PrestaShop and they miss functions, what older version contains. It's very difficult to explain this to customers, why fuction get removed... :(

@web-plus  what do you think? Can you write us your opinion like offical PrestaShop theme developer?

**Comment by web-plus** [comment]
Please keep this feature. I understand the effort to "clean Prestashop", but this feature is really useful and may save quite a lot of developing hours and more important, overall performance of Prestashop shops, especially related to google metrics. Let me explain more in detail.

Google metrics are very strict for mobiles. Loading Javascripts blocks rendering of page and make the metrics bad, so one of the optimization goals is to reduce the amount of JS executed in the page. Most shops need more complex themes then the default theme and also often use many other modules (also often not so optimized). Because of this, I see many PS shops having very bad mobile metrics. When I try to improve mobile google metrics, I often disable modules that are not required on mobiles. This is quick and effective tool, that helps metrics. For example modules from left column that are not visible on mobiles anyway, can be disabled. The feature of disabling module on mobile stops loading the module entirely, which helps a lot (SQL queries, JS libraries...). 

I also develop my own themes and I try to make them OK for google metrics. In some of my themes, I have "Facebook Fanbox widget" module in left column, that loads heavy SDK library and can cause -10 points on mobile metrics. I don't think it is an important module, but other themes has it and you want to be competitive. It is just an example. I am using this code in module:
 $this->disableDevice(Context::DEVICE_MOBILE);
so it is optimized directly after theme installation by disabling the module for mobiles. I use it in more modules (sidebar slider etc.).
Another example may be disabling the homepage slider for mobiles only and display simple webp image instead. 
Or optimizing product mobile page by disabling the "Products from same category" module on shop with lots of categories.
Or, I have clients who want different image size of slider on mobile (square) and PC (wide). I use a clone of slider module for that and I can disable one module on mobile only. Yes, I can rewrite this specific module to have two images in one module, but as I say, this is just an example and it will require extra dev hours for something that was easy before. 
In practice, I can't really see optimizing all modules on all shops to achieve the same effect. I would rather not even offer such service (improving metrics).

I also saw other developers use this in tpl files:
{if Context::getContext()->getDevice() == 1}

Also, having the same experience on mobile as on PC is not something strict from my point of view. Some large shops have dedicated mobile versions etc.

I hope this makes sense and you can reconsider it again.

Jiri

**Comment by ShaiMagal** [comment]
@web-plus Jiri, PS9 will be your nightmare with this PR :D Are your themes already ready for this PR? How you will solve this, when there is no poasibility in PS9 to disable/enable module on mobile device?

**Comment by web-plus** [comment]
I was hoping Prestashop team will at least reply to my previous post :-(

**Comment by kpodemski** [comment]
@web-plus 

Hello Jiri

I think what is most important is that you are free to implement similar feature to your modules, nothing is blocking you from doing that. 

Given that it has been used only with a promile of modules and it's been proven as unreliable (because of the cache issues), I think that was the right thing to do to remove it.

**Comment by web-plus** [comment]
@kpodemski  Thank you for response. My modules in my themes is one thing, trying to optimize metrics for clients that are using many 3rd party modules and different theme is another thing. It would be harder now to optimize Prestashop sites for mobile metrics. Btw, never experienced caching issues for what I am using (disable modules on mobiles). 
To understand us more - @ShaiMagal and me are trying to keep our clients Prestashop sites fast and optimized, using good hosting and optimize our modules and themes. This feature was quite important part of the process, see my previous explanations and the one below. 

@ShaiMagal - I will test / consider using something like https://packagist.org/packages/mobiledetect/mobiledetectlib in my themes or just dont use plugins like Facebook Page Plugin in my themes at all. It will still be a problem because you cant for example just disable modules like "Products from same category" to optimize product page for mobiles only (PC metrics are not strict but you will loose the option to use it on PC only - without modifying all modules). Also, manual changes in default and 3rd party modules will be always rewritten by updates, so I would probably just not offer the optimization service anymore...

**Comment by web-plus** [comment]
@kpodemski - To understand us more - @ShaiMagal and me are trying to keep our clients Prestashop sites fast and optimized,  by using good hosting and optimize our modules and themes. This feature was quite important part of the process...

**Comment by Hlavtox** [comment]
@web-plus MobileDetect is included in PrestaShop already, that how this feature used to work. If you really want to use this way, you can do this:

```php
use Detection\MobileDetect;

$mobileDetect = new MobileDetect();
$is_mobile = $mobileDetect->isMobile();
$is_tablet = $mobileDetect->isTablet();
```

------------------

Jirko, I know that you are pissed about this, but try to look on it from a different angle.

**Problem 1 - All the clients should have the same experience.**
Why hide something on a phone or a tablet? All users should have the same benefits, same functionality, regardless of the device they view the page on. 60% traffic is from mobile devices, why they should not see something that the desktop user does?

**Problem 2 - Precision of such detection**
If I use `$mobileDetect->isMobile();`, what breakpoint does it use? How does it know if my left column is hidden on 768 px or 991 px?
If you use it for hiding something in left column, when I rotate my tablet to landscape, why is it still hidden?
If I use my tablet in landscape mode with a keyboard, is it a desktop or is it a tablet?

-----------------------

And regarding the use cases you mentioned.

> Google metrics are very strict for mobiles. Loading Javascripts blocks rendering of page and make the metrics bad, so one of the optimization goals is to reduce the amount of JS executed in the page. Most shops need more complex themes then the default theme and also often use many other modules (also often not so optimized). Because of this, I see many PS shops having very bad mobile metrics. When I try to improve mobile google metrics, I often disable modules that are not required on mobiles. This is quick and effective tool, that helps metrics.

Reducing amount of JS execution on the page, optimizing it and lazyloading is not related to a mobile, it should be done in all cases. My shops in a category, whole js bundle has 112 kB. Category with 2749 products - https://www.trendo.cz/94-granitove-drezy.

Snímek obrazovky 2024-08-05 104734

> For example modules from left column that are not visible on mobiles anyway, can be disabled. The feature of disabling module on mobile stops loading the module entirely, which helps a lot (SQL queries, JS libraries...).

If the module takes anything more than few ms, it's badly written.

> I also develop my own themes and I try to make them OK for google metrics. In some of my themes, I have "Facebook Fanbox widget" module in left column, that loads heavy SDK library and can cause -10 points on mobile metrics. I don't think it is an important module, but other themes has it and you want to be competitive.

This is done by deferring the script so it doesn't block page loading. If it still slows down the page, you can make a tiny bit of JS that will inject that widget if the left column is visible. You will speed up the desktop loading also, win win situation.

> Another example may be disabling the homepage slider for mobiles only and display simple webp image instead.

Why should mobile users not see the slider? And if you really want to hide it, slider takes virtually 10 lines of HTML and 1 db query. If it's hidden by CSS, the browser won't load the images.

> Or optimizing product mobile page by disabling the "Products from same category" module on shop with lots of categories.

You should optimize for all clients, not just mobile.

> Or, I have clients who want different image size of slider on mobile (square) and PC (wide).

Study srcsets - https://web.dev/patterns/web-vitals-patterns/images/responsive-images - you can serve different aspect ratio images for different resolutions.

**Comment by web-plus** [comment]
@Hlavtox I know your shop trendo and your work, so I know you can optimize shops very well and I think I can do it too. My point was - this tool was enabling to optimize websites on **3rd party modules / default modules** for mobile metrics **fast**. If you take care about hundreds of PS clients, who are purchasing different modules etc, you will never be able to achieve the same quality as with your custom shop. Many clients would not even pay dev time for that (to modify all 3rd party modules etc.). This is why this feature was useful for QUICK performance optimizations. I agree with all you wrote, but I think you guys don't see my point :-) I am not "pissed", just trying to show other perspective (developer who maintains other shops, not developing for core or strictly only my products). 

I still don't think the mobile version must be strictly same, it is quite normal that you try to simplify the mobile version, which is what this feature did well too. But the point here is **not to load the JS libraries on mobiles easily.** The experience may be same or similar, but if I can avoid loading JS libraries on mobile, it is very useful for the metrics which are very strict for mobiles. Do you agree here? E.g. regarding slider, I already use on some website CSS mobile slider (no JS), see ravina.cz. But this mobiledetect feature was great because I could use Slick or OwlCarousel for PC only and disable the library completely for mobile (and use pure CSS only for sliding). For PS9 I will have to add the mobiledetect back, or any other suggestion how to avoid loading the libraries on mobile only? 
I can probably write my own minimalistic JS for everything, but again - more work and I cant optimize all the other 3rd party modules on websites of my clients.

I think we are now both repeating our own arguments. I understand your point of view and appreciate the effort to clean the core code. I welcome all the other work and improvements you do (you are one of the most active contributors and I really appreciate your work). I just don't agree with this one :-) as it will make the life harder for me and @ShaiMagal and probably others. 
If you can think about maybe replacing the feature with something similar, more reliable / clean (to achieve your goals why you do that...).  I personally wouldn't need to use it in tpl files etc. Just be able to disable JS libraries for mobile only basically. 
But if you still decide just to remove it, OK... I just wanted to explain my reasons why I will really miss this feature. 

Jiri

**Comment by ShaiMagal** [comment]
Our eshop owners are using this feature -> and again, if we make update PS for them, they will complain again something like:
"We updated to never PS and we miss feature XYZ" - I already had this problems before when we updated from 1.6 to 1.7 and there were a lot of things missing. (I mean something like 1.7.3.0 etc)

But main goal is @Hlavtox removed working feature and I don't know why. For me and @web-plus it was usefull.

And I have real use case from eshop owners as well.

Image for example any banner or facebook bar on the left side. Eshop owner can to show this on the bigger screens only, so he disable it for "mobile" (smaller screen) to don't take space on screen.

It's real use case. How the eshop owner will do this in PS9? He will must ask programmer and pay him?
Why? We must make PrestaShop friendly for eshop owners, he must be able to do most of things alone (ideally everything BASIC like is hiding module on mobile)

We tell "mobile", but imagine it like "small screen".

Using CSS is not ideal and it's not solution, if the module is injecting any javascript for example, it will be fired and it is useless, it must be blocked totally.

**This PR will complicate life of eshop owners and me itself as well (people, who use this feature, will complain).**

And recommendation of implementing MobileDetect to every single moduly is crazy...
Because we need to disable on "mobile" any 3rd module or native PrestaShop module, and we don't want to edit core of every module or make any custom changes. Waste of human resources.

@Hlavtox make great job for PrestaShop community, a lot of PR! :) but sorry bro, this PR is blunder.

**Comment by ShaiMagal** [comment]
@kpodemski 
You wrote:
_I think what is most important is that you are free to implement similar feature to your modules, nothing is blocking you from doing that._


We aren't talking only about our modules, but we talk globally, we need to "hide" on "small screens" 3rd modules as well, so implementing feature to every module is insanity....

Before this PR, there were everything, what we need and @Hlavtox removed it.

Btw, this type of PR is one of the more reasons why people are leaving from PrestaShop to Shoptet... (in Czech republic at least)

So, please, don't complicate life of eshop owners and developer who cares about perfomance like is me and @web-plus for example.

**Comment by kpodemski** [comment]
@ShaiMagal 

In over 15 years working with PrestaShop I've used this option once when I wanted to disable Revolution Slider. Nowadays, with sourceset any many other solutions it's hard to imagine a need to differentiate between mobile and desktop, so that you don't load something what supposedly is heavy for the performance and there is no alternative.

I feel like you are searching for a problem where there aren't any. 

Could you provide scenarios where you have used this option and there was no alternative? I'm really curious about real-life use cases you have in mind.

**Comment by kpodemski** [comment]
@ShaiMagal 

I just read examples you shared with Daniel and they are just like I imagined, any of them can be done without the removed option.

Eventually, we did what we believe is right. Maybe, if there will be lot more reports, we can bring the feature back.

**Comment by Hlavtox** [comment]
@ShaiMagal We are really repeating ourselves here. If you change your mindset, you will see the benefits in the long run.

@integritty did not see any issue with this removed.

> And I have real use case from eshop owners as well.
> 
> Image for example any banner or facebook bar on the left side. Eshop owner can to show this on the bigger screens only, so he disable it for "mobile" (smaller screen) to don't take space on screen.
> 
> It's real use case. How the eshop owner will do this in PS9? He will must ask programmer and pay him?
> Why? We must make PrestaShop friendly for eshop owners, he must be able to do most of things alone (ideally everything BASIC like is hiding module on mobile)
> 
> We tell "mobile", but imagine it like "small screen".
> 
> Using CSS is not ideal and it's not solution, if the module is injecting any javascript for example, it will be fired and it is useless, it must be blocked totally.

As I said, so this widget is slow and is ruining your performance on page load right?
So if you hide this on mobiles using this feature, it's still ruining the performance on desktop right? You don't see this as a problem? Inject this lazily after page load with javascript and you will benefit ALL screens and it will work properly.

Please, provide arguments if you think this is not true. :-)

**Comment by web-plus** [comment]
@Hlavtox @kpodemski 
You asked for arguments but didn't reply to some of my arguments in last post. I am sorry, I will just one last time try to summarize my point of view (which is similar with @ShaiMagal - we work together quite a lot recently, having the same goal - keep shops of our clients fast and with OK google metrics - basically without clients paying something extra for it, to stay competitive. I basically agree with your opinions and point of view, but we all work on different things, providing different services and that may change the point of view. So these are our arguments / reasons why we liked this feature. 

Imagine you take care of hundreds of shops and want to keep them with relatively good google metrics for mobiles. Take into consideration that only mobile metrics are much more strict. Take into consideration thousands of modules from addons, often not optimized for performance (loading massive JS like mentioned Revolution slider - see number of downloads. You cant write your own modules for everything or customize everything. And if, why use Prestashop that was supposed to be quite cheap to stay competitive? When working on shop with modules from 3rd party developers, all these modules needs to be updated in time too, so custom modifications are not smart here. 

I believe you can't optimize everything with reasonable costs or inject content with JS always. Btw this method (inject via JS) hurt metrics too in some cases - CLS ( you may not know height of the element to reserve it and avoid CLS). Also performance - test it with and without default Product comments module, which injects content via JS too (we discussed this before, I was again argumenting with metrics and issues with positioning of elements that comes with this method - you explained that doing it with hooks would move performance issues to DB, which I understand and I was OK with it). It is modern to do everything via JS (React, Vue), but injecting too many content to DOM is not "google metrics friendly". I dont want to be slave to metrics, but you cant just ignore them too and some clients check them too. 

You can remove the feature, I really don't push you, but shops of our clients will have worse mobile metrics, clients will simply not pay for complex optimizations. Our clients often understand Prestashop as "not so expansive" platform, but they still expects quite high standards. Often we work on shops that we didn't create, so the quality is not top as you reference with your dedicated projects. And if we upgrade shop for client, some of them test metrics and see it is bad after installing their needed xy 3rd party modules. They are not happy to hear that custom optimization will be expensive. This is where the feature was fast and effective, but I think not many developers knows about it, I came to it not long ago too. I admit it is rather creative way and we can talk hours about best dev approach etc. 

This is related topic actually - Prestashop gets more and more expensive due to quite heigh minimum price of addons on marketplace (you cant offer some easy solutions cheaply on addons anymore), all the subscriptions, raising dev prices. Sometimes it just doesn't make sense for certain shops and I understand why they migrate to cloud cheap services. On our market, Prestashop used to be considered as quite cheap option to run the shop, but this is changing in last years and the environment got really competitive too (cloud services - Shoptet being the biggest one).

**Comment by kpodemski** [comment]
> And if, why use Prestashop that was supposed to be quite cheap to stay competitive?

To be honest, I wanted to stop reading after that. Why do you think that PrestaShop supposed to be a "quite cheap solution"? PrestaShop is not a cheap solution, and it is not for merchants who don't have enough money to maintain the software. I feel like there's still a myth in our ecosystem that PrestaShop aims to be cheap and affordable for all kinds of merchants, even the smallest ones. I personally don't think that's the case anymore :)

> I believe you can't optimize everything with reasonable costs or inject content with JS always. Btw this method (inject via JS) hurt metrics too in some cases - CLS ( you may not know height of the element to reserve it and avoid CLS). Also performance - test it with and without default Product comments module, which injects content via JS too (we discussed this before, I was again argumenting with metrics and issues with positioning of elements that comes with this method - you explained that doing it with hooks would move performance issues to DB, which I understand and I was OK with it). It is modern to do everything via JS (React, Vue), but injecting too many content to DOM is not "google metrics friendly". I dont want to be slave to metrics, but you cant just ignore them too and some clients check them too.

You can optimize everything, but it does cost money. How would disabling the product comments module on mobile help in this scenario? That's not a solution. It's a workaround.

> This is related topic actually - Prestashop gets more and more expensive due to quite heigh minimum price of addons on marketplace (you cant offer some easy solutions cheaply on addons anymore), all the subscriptions, raising dev prices. Sometimes it just doesn't make sense for certain shops and I understand why they migrate to cloud cheap services. On our market, Prestashop used to be considered as quite cheap option to run the shop, but this is changing in last years and the environment got really competitive too (cloud services - Shoptet being the biggest one).

I think that's the main point of this discussion. You think the PrestaShop project should focus on smaller merchants and stay "not expensive", but I personally think we should focus on developers and bigger merchants and provide them with the tools to create the best e-commerce possible, not all the features in the world, but freedom to create those, and a good foundation.

To be honest, during the time you guys spent on this discussion, you could've created a module that would bring back this solution. It is as simple as having one module where you select which module you don't want to display on mobile, and possibly you override the `Hook:: getAllHookRegistrations ` method. That's what PrestaShop is all about: you have the freedom to do things like you want.

Maybe there is some misunderstanding somewhere? The `mobile_theme` feature hasn't been used since PrestaShop 1.5, if I recall correctly, so the only change in this PR is that you no longer have a button to disable the module on mobile, a feature that is quite straightforward to re-implement using a custom module (one custom module). The tools to check whether you are on mobile devices or not are still there. 

I wanted to emphasize that the above is my personal opinion, not an official reply from PrestaShop SA or so (I'm not a PrestaShop SA employee anyway, but I wanted to make it clear).

**Comment by web-plus** [comment]
Btw, for my own modules in themes, I was thinking about using something simple like

`
$userAgent = $_SERVER['HTTP_USER_AGENT'];
    $isMobile = preg_match('/Mobile|Android|BlackBerry|IEMobile|Silk|Opera Mini|webOS|Windows Phone|iPhone|iPad|iPod|Symbian|BB10|PlayBook|Nokia|Samsung|Palm|Fennec|Bada|Maemo|Tizen|MeeGo|UC Browser|KAIOS/i', $userAgent);

    if (!$isMobile) {
"load JS library"
`

Maybe if we find some way to do something similar in core, without the library (if thats what bothers you) to keep that "Disable for mobile" button working (the module will not load at all, mobile version will save SQL queries, avoid JS, etc.), we can all be happy?

**Comment by web-plus** [comment]
> To be honest, I wanted to stop reading after that. Why do you think that PrestaShop supposed to be a "quite cheap solution"? PrestaShop is not a cheap solution, and it is not for merchants who don't have enough money to maintain the software. I feel like there's still a myth in our ecosystem that PrestaShop aims to be cheap and affordable for all kinds of merchants, even the smallest ones. I personally don't think that's the case anymore :)

OK, this explains our different opinions and approach too. I think such discussions is not waste of time, I understand you more now. We have some clients with very big buget too, but many smaller ones, so we have to work with that. We see some of them migrating to cloud local service (they actively contact Prestashop merchants specifically and try to migrate them). So this is for context. Btw, I am more into frontend, so it would probably take me more time to come with good solution, so I just tried to discuss it and maybe come up with something together, which would work as a standard maybe. But we can probably find our solution with @ShaiMagal. The module that can be installed and bring the functionality back, is a good idea.

Thank you for your time (I mean it).

Jiri

**Pull request merged**